### ISSUE TITLE
Slowness compared to rstanarm

### ISSUE THREAD

**Issue by waynefoltaERI** [paul-buerkner/brms#24]
I love `brms`, and am currently writing a blog post about it. Just discovered `rstanarm`, which is similar, but `brms` is better in most every case. The one thing that it lacks is that it seems to be slower to run. When I do a logistic regression on the last two Iris classifications, `rstanarm` runs in about 7 or 8 seconds from the time I hit return, while `brms` takes 30-50 seconds.

Not a show-stopper, and I imagine the gap is mainly due to fixed-cost overhead, so matters less and less on larger problems, but it's noticeable. Also, if both `brms` and `rstanarm` are loaded, `brm` also appears to run a bit more slowly than it does when that's not the case. (The latter isn't an important use case, except that folks might have both loaded when comparing them.)

**Comment by paul-buerkner** [comment]
This large speed gap is strange. Can you provide a replicable example? In most tests I have done so far, `brms` and `rstanarm` had very similar speed, while `brms` was usually slightly faster.

**Comment by paul-buerkner** [comment]
Edit: Did you include the compilation time? Because that would explain the difference. 

The reason is that `brms` writes all Stan models from scratch and has to compile them, while `rstanarm` comes with precompiled code. The advantage of the `brms` approach is that the stan code is easier to write and read. Also it may be slightly faster after having compiled the model. In this sence, you are right that this is a fixed cost overhead. Furthermore, precompilation of Stan code was practically impossible when I started to build `brms` and it only became possible recently with the relase of `rstan 2.8.2`.

The advantage of the `rstanarm` approach is obviously that no compilation is needed anymore, thus saving the compilation time and the need for a C++ compiler

**Comment by waynefoltaERI** [comment]
```R
library (caret)
library (brms)

# load ("~/Documents/Blog/brms blog post/brms blog.RData")

# devtools::install_github ("paul-buerkner/brms")

rstan_options (auto_write=TRUE)
options (mc.cores=parallel::detectCores ())

set.seed (3875)

ir <- data.frame (scale (iris[, -5]), Species=iris[, 5])

ir2 <- iris[-c(1:50),]
ir2$Species <- factor (ir2$Species)

library (rstanarm)  # Overrides cauchy, fixef, ngrps, ranef, and VarCorr from package brms!!
library (rstan)
library (shinystan)

### Outputs:

Compiling the C++ model
starting worker pid=5769 on localhost:11823 at 14:18:30.208
starting worker pid=5777 on localhost:11823 at 14:18:30.394
starting worker pid=5785 on localhost:11823 at 14:18:30.582

SAMPLING FOR MODEL 'bernoulli(logit) brms-model' NOW (CHAIN 1).

Chain 1, Iteration:    1 / 3000 [  0%]  (Warmup)
SAMPLING FOR MODEL 'bernoulli(logit) brms-model' NOW (CHAIN 2).

Chain 2, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 1, Iteration:  300 / 3000 [ 10%]  (Warmup)
SAMPLING FOR MODEL 'bernoulli(logit) brms-model' NOW (CHAIN 3).

Chain 3, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 1, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 1, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 2, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 2, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 2, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 3, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 1, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 3, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 3, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 2, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 1, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 3, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 1, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 2, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 3, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 1, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 2, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 3, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 1, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 2, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 3, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 1, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 2, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 3, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 1, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 2, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 1, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.61468 seconds (Warm-up)
#                5.97053 seconds (Sampling)
#                7.58522 seconds (Total)
# 

Chain 2, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.51921 seconds (Warm-up)
#                6.04108 seconds (Sampling)
#                7.56029 seconds (Total)
# 

Chain 2, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.5344 seconds (Warm-up)
#                6.77796 seconds (Sampling)
#                8.31236 seconds (Total)
# 
   user  system elapsed 
 24.035   1.209  45.316 

system.time (br2 <- stan_glm (Species ~ Petal.Length + Petal.Width + Sepal.Length + Sepal.Width, data=ir2,
                              family=binomial (link="logit"), chains=3, iter=3000, warmup=600,
                              prior=normal (0, 5), prior_intercept=normal (0, 5)))

### Outputs:

starting worker pid=5805 on localhost:11823 at 14:19:23.406
starting worker pid=5813 on localhost:11823 at 14:19:23.608
starting worker pid=5821 on localhost:11823 at 14:19:23.796

SAMPLING FOR MODEL 'bernoulli' NOW (CHAIN 1).

Chain 1, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 1, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 1, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 1, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 1, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 1, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 1, Iteration: 1500 / 3000 [ 50%]  (Sampling)
SAMPLING FOR MODEL 'bernoulli' NOW (CHAIN 2).

Chain 2, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 1, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 2, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 1, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 2, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 2, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 1, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 2, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 1, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 2, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 1, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 0.228873 seconds (Warm-up)
#                0.838139 seconds (Sampling)
#                1.06701 seconds (Total)
# 

SAMPLING FOR MODEL 'bernoulli' NOW (CHAIN 3).

Chain 3, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 2, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 3, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 2, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 3, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 3, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 2, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 2, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 2, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 2, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 0.257968 seconds (Warm-up)
#                0.885888 seconds (Sampling)
#                1.14386 seconds (Total)
# 

Chain 3, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 3, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 3, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 3, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 0.236465 seconds (Warm-up)
#                1.08039 seconds (Sampling)
#                1.31686 seconds (Total)
# 
   user  system elapsed 
  0.407   0.115   7.477 
```
Note that there's a long delay after your "Compiling the C++ model" that doesn't exist in the `rstanarm` version. (The `rstanarm` version immediately prints the multiple processes starting message.) This could be due to the `rstan_options (auto_write=TRUE)`, and perhaps reusing a model. Not sure.

Again, it seems to be fixed overhead, mainly tied in to compiling the model. Though the iteration messages seem to come out about twice as fast -- just guessing here as they both move fairly fast -- for `rstanarm` as for `brms`.

Does that help?

**Comment by paul-buerkner** [comment]
That helps, thanks alot! I don't see the call to the `brm` function, though. Can you also provide it?

**Comment by waynefoltaERI** [comment]
Okay, updated the previous to include the `brms` call. Sorry about that.

**Comment by paul-buerkner** [comment]
You are right, for this model it appears to be slower even after taking compilation time into account. My first guess is that this is because of the implementation of the bernoulli distribution in Stan. I will see, if I can improve the speed for this type of model.

Would you mind testing the speed of both packages also for some examples using other families? I am interested if your results match mine in those cases (that is speed is overall very similar when ignoring compilation time).

**Comment by waynefoltaERI** [comment]
When I try this:

```R
> system.time (foo1 <- brm (mpg ~ ., data=mtcars, prior=c(set_prior ("normal (0, 7)", class="b")), n.chains=3, n.iter=3000, n.warmup=600))
Compiling the C++ model
starting worker pid=5991 on localhost:11823 at 14:42:09.352
starting worker pid=5999 on localhost:11823 at 14:42:09.535
starting worker pid=6007 on localhost:11823 at 14:42:09.728

SAMPLING FOR MODEL 'gaussian(identity) brms-model' NOW (CHAIN 1).

Chain 1, Iteration:    1 / 3000 [  0%]  (Warmup)
SAMPLING FOR MODEL 'gaussian(identity) brms-model' NOW (CHAIN 2).

Chain 2, Iteration:    1 / 3000 [  0%]  (Warmup)
SAMPLING FOR MODEL 'gaussian(identity) brms-model' NOW (CHAIN 3).

Chain 3, Iteration:    1 / 3000 [  0%]  (Warmup)
Chain 1, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 1, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 1, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 2, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 2, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 2, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 1, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 3, Iteration:  300 / 3000 [ 10%]  (Warmup)
Chain 1, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 2, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 3, Iteration:  600 / 3000 [ 20%]  (Warmup)
Chain 3, Iteration:  601 / 3000 [ 20%]  (Sampling)
Chain 1, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 2, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 1, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 3, Iteration:  900 / 3000 [ 30%]  (Sampling)
Chain 2, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 1, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 3, Iteration: 1200 / 3000 [ 40%]  (Sampling)
Chain 2, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 1, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration: 1500 / 3000 [ 50%]  (Sampling)
Chain 2, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 1, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 2, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration: 1800 / 3000 [ 60%]  (Sampling)
Chain 1, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.92806 seconds (Warm-up)
#                3.44862 seconds (Sampling)
#                5.37668 seconds (Total)
# 

Chain 2, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 2100 / 3000 [ 70%]  (Sampling)
Chain 2, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.61481 seconds (Warm-up)
#                3.86154 seconds (Sampling)
#                5.47635 seconds (Total)
# 

Chain 3, Iteration: 2400 / 3000 [ 80%]  (Sampling)
Chain 3, Iteration: 2700 / 3000 [ 90%]  (Sampling)
Chain 3, Iteration: 3000 / 3000 [100%]  (Sampling)# 
#  Elapsed Time: 1.6777 seconds (Warm-up)
#                4.15795 seconds (Sampling)
#                5.83566 seconds (Total)
# 
   user  system elapsed 
 23.545   1.091  42.310 
```
It takes 35 seconds from hitting enter until seeing the first iteration message. So after that it runs in about 7 seconds. The `stan_glm` takes 8 seconds, but also seems to have less delay between printing the multiple-threads-starting messages and actually outputting the Iteration messages.

Ahh, I'm nearly certain that `rstanarm` uses `Rcpp`, and maybe it either tells `rstan` to bypass clang and use `Rcpp` instead, or it bypasses `rstan` completely and uses `Rcpp`. Just trying to guess how your compile takes 35 seconds -- which I seem to remember is normal for direct `rstan` usage -- versus `rstanarm`'s near-instantaneous compilation. That kind of trickery may not be worth it. (I believe the `rstanarm` people are also the Stan and `rstan` people, so they may pull tricks that a third party can't.)

It's mostly only an issue with toy problems. (I'm on vacation and don't have access to work projects now, so I only have toy problems to play with.)

**Comment by jgabry** [comment]
On Mon, Jan 11, 2016 at 2:49 PM, waynefoltaERI <notifications@github.com>
wrote:

> Ahh, I'm nearly certain that rstanarm uses Rcpp, and maybe it either
> tells rstan to bypass clang and use Rcpp instead, or it bypasses rstan
> completely and uses Rcpp. Just trying to guess how your compile takes 35
> seconds -- which I seem to remember is normal for direct rstan usage --
> versus rstanarm's near-instantaneous compilation.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/paul-buerkner/brms/issues/24#issuecomment-170669039>.
>
@waynefoltaERI To follow up on Paul's response, rstanarm comes with
everything already compiled. The Stan code is written to allow for all of
the various options you can specify when calling the rstanarm modeling
functions (e.g. stan_glm) using a bunch of conditional logic. So the models
can all be compiled once (nd only once) when the package builds and there
should be zero compilation time when using the package.

On the other hand, brms takes the approach of writing the Stan code for you
each time you specify a model. So you have to wait for compilation, but
this approach does allow for additional flexibility beyond what rstanarm is
intended to do.

**Comment by waynefoltaERI** [comment]
Good explanation. I much prefer the `brms` approach. Just good to have a written explanation of this particular difference. Maybe I just missed it in the documentation, but if it's not there it would be nice to add. (Especially since lots of folks might compare `brms` to the "name brand" `rstanarm`.)

I've found `brms` to be more flexible and to have fewer issues than `rstanarm` by a long shot. The only two things that `rstanarm` has at this point are: 1) faster run on smaller problems -- though this has an inflexibility downside -- and 2) GAMM capability. Other than that, it's way less flexible and reliable, in my experience so far.

(For example, while playing with the `mtcars` dataset for this issue, I found that `brms`' and `rstanarm`'s answers differed considerably. And `glm` agreed with `rstanarm`... which gave me a clue and when I investigated further, it appears that `rstanarm` is ignoring priors. (Or I am misunderstanding how they are specified, but not getting any error messages.)  So the reason for the agreement is that I was specifying priors, but `rstanarm` was ignoring them and using flat (improper, frequentist-like) priors.)

I can

**Comment by waynefoltaERI** [comment]
Issue resolves to possibly updating documentation. Closing.

**Comment by jgabry** [comment]
@waynefoltaERI I just opened an rstanarm issue [here](https://github.com/stan-dev/rstanarm/issues/48). Would you mind responding there so we can discuss your mtcars example? Thanks!

**Comment by paul-buerkner** [comment]
I also was a bit confused on the differing results and I am not sure if it is really a problem in `rstanarm`.
I rather think that there might be a problem with the bernoulli models in `brms`. I will investigate this further.

**Comment by paul-buerkner** [comment]
I guess the differences in the results are a good example of why multicollinearity is bad for regression models: 

```R
cor(ir2[, 1:4])
             Sepal.Length Sepal.Width Petal.Length Petal.Width
Sepal.Length    1.0000000   0.5538548    0.8284787   0.5937094
Sepal.Width     0.5538548   1.0000000    0.5198023   0.5662025
Petal.Length    0.8284787   0.5198023    1.0000000   0.8233476
Petal.Width     0.5937094   0.5662025    0.8233476   1.0000000

library(car)
ir2$y <- rnorm(nrow(ir2))
fit <- lm(y ~ Petal.Length + Petal.Width + Sepal.Length + Sepal.Width, data = ir2)
vif(fit) # variance inflation factors
Petal.Length  Petal.Width Sepal.Length  Sepal.Width 
    7.252447     3.948354     3.990113     1.721954 
```

When I run
```R
br1 <- brm(Species ~ Petal.Length + Petal.Width, data=ir2,
           family=bernoulli(link="logit"), chains=3, iter=3000, warmup=600,
           prior = c(set_prior("normal(0,10)"), 
                     set_prior("normal(0,50)", coef = "Intercept")))
br1
## output
Family: bernoulli (logit) 
Formula: Species ~ Petal.Length + Petal.Width 
   Data: ir2 (Number of observations: 100) 
Samples: 3 chains, each with iter = 3000; warmup = 600; thin = 1; 
         total post-warmup samples = 7200
   WAIC: 26.28
 
Fixed Effects: 
             Estimate Est.Error l-95% CI u-95% CI Eff.Sample Rhat
Intercept      -46.33     12.13   -73.43   -26.27       1289    1
Petal.Length     6.13      2.21     2.34    11.12       1361    1
Petal.Width      9.99      3.33     4.20    17.11       1888    1

Samples were drawn using sampling(NUTS). For each parameter, Eff.Sample 
is a crude measure of effective sample size, and Rhat is the potential 
scale reduction factor on split chains (at convergence, Rhat = 1).

br2 <- stan_glm(Species ~ Petal.Length + Petal.Width, data=ir2,
                family=binomial(link="logit"), chains=3, iter=3000, warmup=600,
                prior_intercept = normal(0, 50), prior = normal(0, 10),
                prior_ops = prior_options(scaled = FALSE))
br2
## output
stan_glm(formula = Species ~ Petal.Length + Petal.Width, family = binomial(link = "logit"), 
    data = ir2, chains = 3, iter = 3000, warmup = 600, prior = normal(0, 
        10), prior_intercept = normal(0, 50), prior_ops = prior_options(scaled = FALSE))

Estimates:
             Median MAD_SD
(Intercept)  -48.3   13.0 
Petal.Length   6.4    2.4 
Petal.Width   10.3    3.5 

Sample avg. posterior predictive 
distribution of y (X = xbar):
         Median MAD_SD
mean_PPD 0.5    0.0  

br3 <- glm(Species2 ~ Petal.Length + Petal.Width, data=ir2,
           family=binomial(link="logit"))
summary(br3)
## output
Call:
glm(formula = Species2 ~ Petal.Length + Petal.Width, family = binomial(link = "logit"), 
    data = ir2)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-1.73752  -0.04749  -0.00011   0.02274   1.89659  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)    
(Intercept)   -45.272     13.610  -3.327 0.000879 ***
Petal.Length    5.755      2.306   2.496 0.012565 *  
Petal.Width    10.447      3.755   2.782 0.005405 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 138.629  on 99  degrees of freedom
Residual deviance:  20.564  on 97  degrees of freedom
AIC: 26.564

Number of Fisher Scoring iterations: 8
```
all three models produce very similar results (at least on my machine). Using all 4 predictors also leads to similar results when using the above priors. However, when I remove the prior specifications in the `brms` model, and thus use flat priors for the regression coefficients, we get those weird differences, we both stumbled upon.

**Comment by paul-buerkner** [comment]
Despite all this, it appears to me that `rstanarm` is faster then `brms` when fitting fixed effects only. I will open a separate issue for it and see if I can improve the sampling speed for these models.
